### PR TITLE
Fix build on macOS with recent clang

### PR DESCRIPTION
**Pull request by jcelerier**
Fix building with clang due to std::is_literal_type being removed

### CODE REVIEW

**Comment by ned14** [comment]
Could not the MSVC and clang preprocessor branches be combined?

**Comment by BurningEnlightenment** [comment]
Shouldn't this also switch on a stdlib selection macro? E.g. GCC with libc++ would still not work, right?

**Comment by ned14** [comment]
> Shouldn't this also switch on a stdlib selection macro? E.g. GCC with libc++ would still not work, right?

The fix is calling the compiler intrinsic, so switching on compiler not standard library seems the right call?

**Comment by jcelerier** [comment]
> The fix is calling the compiler intrinsic

i can confirm this

File: test/main.cpp [hunk 1]
```diff
@@ -54,20 +54,19 @@ Distributed under the Boost Software License, Version 1.0.
...
+  static constexpr bool value = __is_literal_type(T);
+};
+#elif defined(__GNUC__)
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 template <class T> struct is_literal_type
 {
   static constexpr bool value = std::is_literal_type<T>::value;
 };
 #pragma GCC diagnostic pop
```
Comment (BurningEnlightenment) [inline test/main.cpp:69]: I referred to this part. If `GCC` with `libc++` is used `defined(__clang__)` should be false, and we would try to use the non-existant `std::is_literal_type<T>::value` or am I missing something?

Comment (jcelerier) [reply test/main.cpp:69]: I just checked and libstdc++'s source code actually calls GCC's intrinsic which is... exactly the same form than the others. So maybe we can just use __is_literal_type(T) everywhere

Comment (jcelerier) [reply test/main.cpp]: okay, I just went through godbolt and tried every compiler in there, EDG, ICC even the random TI DSP thing, and they *all* support __is_literal_type(T) so I'd suggest just using this no matter what.

**Review by BurningEnlightenment — APPROVED** [review]
Awesome! Thanks for digging into this!

**Review by ned14 — APPROVED** [review]

**Pull request merged**